modules/level_set/src/base/LevelSetProblem.C

```diff
@@ -23,35 +23,54 @@ validParams<LevelSetProblem>()
 
 LevelSetProblem::LevelSetProblem(const InputParameters & parameters) : FEProblem(parameters) {}
 
-void
+bool
 LevelSetProblem::adaptMesh()
 {
+  // reset cycle counter
+  _cycles_completed = 0;
+
   if (!_adaptivity.isAdaptivityDue())
-    return;
+    return false;
 
   unsigned int cycles_per_step = _adaptivity.getCyclesPerStep();
-  _cycles_completed = 0;
-  for (unsigned int i = 0; i < cycles_per_step; ++i)
-  {
 
-    execMultiAppTransfers(LevelSet::EXEC_ADAPT_MESH, MultiAppTransfer::TO_MULTIAPP);
+  Moose::perf_log.push("Adaptivity: adaptMesh()", "Execution");
+
+  bool mesh_changed = false;
 
+  for (unsigned int i = 0; i < cycles_per_step; ++i)
+  {
     _console << "Adaptivity step " << i + 1 << " of " << cycles_per_step << '\n';
+
     // Markers were already computed once by Executioner
     if (_adaptivity.getRecomputeMarkersFlag() && i > 0)
       computeMarkers();
+
+    execMultiAppTransfers(LevelSet::EXEC_ADAPT_MESH, MultiAppTransfer::TO_MULTIAPP);
+
     if (_adaptivity.adaptMesh())
     {
-      meshChanged();
+      mesh_changed = true;
+
+      meshChangedHelper(true); // This may be an intermediate change
       _cycles_completed++;
     }
     else
     {
       _console << "Mesh unchanged, skipping remaining steps..." << std::endl;
-      return;
+      break;
```

Comment:
👍  - This actually fixes a potential bug with matching push/pops with the perf_log.